client/consensus/aura/src/lib.rs

```diff
@@ -264,11 +271,27 @@ impl<B, C, E, I, P, Error, SO> sc_consensus_slots::SimpleSlotWorker<B> for AuraW
 		StorageChanges<sp_api::TransactionFor<C, B>, B>,
 		Self::Claim,
 		Self::EpochData,
-	) -> sp_consensus::BlockImportParams<B, sp_api::TransactionFor<C, B>> + Send> {
-		Box::new(|header, header_hash, body, storage_changes, pair, _epoch| {
+	) -> Result<sp_consensus::BlockImportParams<B, sp_api::TransactionFor<C, B>>, sp_consensus::Error> + Send + 'static> {
+		let keystore = self.keystore.clone();
+		Box::new(move |header, header_hash, body, storage_changes, pair, _epoch| {
 			// sign the pre-sealed hash of the block and then
 			// add it to a digest item.
-			let signature = pair.sign(header_hash.as_ref());
+			let public = pair.public().to_raw_vec();
+			let public_type_pair = pair.public().to_public_crypto_pair();
+			let signature = keystore.read()
+				.sign_with(
```

Comment:
It would be nice if all this could be moved into the keystore trait as a default implementation:
```
fn sign_with_type<P: AppKey>(public: P, data: &[u8]) -> P::Signature {
    // your code
}
```

This makes the code reusable between aura & babe. As this is a default implementation, it should work on the trait.

Reply:
hmm... my initial thought is:
1. BareCryptoStore is defined in sp_crypto
2. AppKey is defined in sp_application_crypto

Do you think we should have this cyclic dependency?

Reply:
I also tried your suggestion with `<P: Public>` and this creates problems:

```
error[E0038]: the trait `traits::BareCryptoStore` cannot be made into an object
   --> primitives/core/src/traits.rs:184:1
    |
54  |   pub trait BareCryptoStore: Send + Sync {
    |             --------------- this trait cannot be made into an object...
...
130 |       fn sign_with_type<P: Public>(
    |          -------------- ...because method `sign_with_type` has generic type parameters
...
184 | / sp_externalities::decl_extension! {
185 | |     /// The keystore extension to register/retrieve from the externalities.
186 | |     pub struct KeystoreExt(BareCryptoStorePtr);
187 | | }
    | |_^ the trait `traits::BareCryptoStore` cannot be made into an object
    |
```

AppKey would have the same problem since it has super traits which are not object safe as well.